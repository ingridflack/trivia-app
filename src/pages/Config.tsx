import { Box, Card, Container, Image, Text } from "@chakra-ui/react";
import TriviaConfig from "../components/TriviaConfig";
import Header from "../components/Header";
import { Footer } from "../components/Footer";

export default function Config() {
  return (
    <>
      <Header />

      <Container
        display="flex"
        alignItems="center"
        justifyContent="center"
        minH="calc(var(--chakra-vh) - 50px)" // 100vh - 50px (height of the header)
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          paddingY="50px"
          paddingX="70px"
          padding={{ base: "20px", md: "50px 70px" }}
          shadow="xl"
        >
          <Image
            src="/src/assets/setup.svg"
            alt="Login image"
            width="388px"
            height="307px"
            pointerEvents="none"
            display={{ base: "none", md: "none", lg: "block" }}
          />
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="20px" color="gray.700">
              Trivia setup
            </Text>
            <TriviaConfig />
          </Box>
        </Card>
      </Container>
      <Footer />
    </>
  );
}
