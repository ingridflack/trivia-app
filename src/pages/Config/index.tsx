import { Box, Card, Container, Image, Text, useToast } from "@chakra-ui/react";
import TriviaConfig from "../../components/TriviaConfig";
import Header from "../../components/Header";
import { Footer } from "../../components/Footer";
import * as UserService from "../../services/userService";
import * as TriviaService from "../../services/triviaService";
import { SubmitHandler, useForm } from "react-hook-form";
import { useNavigate } from "react-router-dom";
import { useState } from "react";
import { MultiValue } from "react-select";
import { UserSearchSelectOption } from "../../types/sharedTypes";
import useAuth from "../../hooks/useAuth";

export interface GameConfigValues {
  category: string;
  difficulty: string;
  amount: number;
  gameMode: string;
  type: string;
}

export default function Config() {
  const {
    register,
    handleSubmit,
    formState: { errors, isSubmitting },
  } = useForm<GameConfigValues>({
    defaultValues: {
      difficulty: "any",
      type: "any",
      amount: 5,
    },
  });

  const [gameMode, setGameMode] = useState("single");
  const navigate = useNavigate();
  const toast = useToast();
  const [invitedUsers, setInvitedUsers] = useState<
    MultiValue<UserSearchSelectOption>
  >([]);
  const { userData } = useAuth();

  const onSubmit: SubmitHandler<GameConfigValues> = async (values) => {
    try {
      const invitedUsersIds =
        gameMode === "multi"
          ? invitedUsers.map((option) => option.value)
          : undefined;

      const body = {
        ...values,
        invitedUsers: invitedUsersIds,
      };

      const { data } = await TriviaService.createTrivia(body);

      navigate(`/trivia/${data.triviaId}`);
    } catch {
      console.log("error");

      toast({
        title: "Oops!",
        description: "Something went wrong. Try again.",
        status: "error",
        duration: 5000,
        isClosable: true,
        position: "top-right",
      });
    }
  };

  const handleGameModeChange = (value: string) => {
    setGameMode(value);
  };

  const handleInviteUser = (value: MultiValue<UserSearchSelectOption>) => {
    setInvitedUsers(value);
  };

  const handleLoadUsers = async (username: string) => {
    const currentUserId = userData ? userData.id : "";

    const { data } = await UserService.search({ username });
    return data
      .filter((user) => user._id !== currentUserId)
      .map((user) => ({ value: user._id, label: user.username }));
  };

  return (
    <>
      <Header />

      <Container
        as="main"
        display="flex"
        alignItems="center"
        justifyContent="center"
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          paddingY="50px"
          paddingX="70px"
          padding={{ base: "20px", md: "50px 70px" }}
          shadow="xl"
        >
          <Image
            src="/assets/setup.svg"
            alt="Login image"
            width="388px"
            height="307px"
            pointerEvents="none"
            display={{ base: "none", md: "none", lg: "block" }}
          />
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="20px" color="gray.700">
              Trivia setup
            </Text>
            <TriviaConfig
              onLoadUsers={handleLoadUsers}
              onInviteUser={handleInviteUser}
              onGameModeChange={handleGameModeChange}
              onSubmit={handleSubmit(onSubmit)}
              errors={errors}
              register={register}
              isSubmitting={isSubmitting}
              gameMode={gameMode}
              invitedUsers={invitedUsers}
            />
          </Box>
        </Card>
      </Container>
      <Footer />
    </>
  );
}
