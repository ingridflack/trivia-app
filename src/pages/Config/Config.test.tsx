import { describe, expect, it } from "vitest";
import Config from ".";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

describe("Config", () => {
  it("should match the snapshot", () => {
    const { container } = render(
      <MemoryRouter>
        <Config />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
