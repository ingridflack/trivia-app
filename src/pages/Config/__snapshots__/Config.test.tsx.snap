// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Config > should match the snapshot 1`] = `
<div>
  <header
    class="css-cyjglw"
  >
    <div
      class="chakra-container css-old1by"
    >
      <div
        class="chakra-stack css-2jbh02"
      >
        <a
          href="/"
        >
          <p
            class="chakra-text css-kkaoep"
          >
            Trivia
          </p>
        </a>
        <div
          class="css-j0ozid"
        >
          <a
            href="/login"
          >
             Sign in 
          </a>
          <a
            href="/sign-up"
          >
             Sign Up 
          </a>
        </div>
      </div>
    </div>
  </header>
  <main
    class="chakra-container css-3i1d8n"
  >
    <div
      class="chakra-card css-11oh4yf"
    >
      <img
        alt="Login image"
        class="chakra-image css-1k0ltxj"
        src="/assets/setup.svg"
      />
      <div
        class="css-1updlva"
      >
        <p
          class="chakra-text css-1yfmrqw"
        >
          Trivia setup
        </p>
        <form
          role="form"
          style="width: 100%; display: flex; align-items: center; justify-content: center; flex-direction: column;"
        >
          <div
            class="css-1ggap6e"
          >
            <div
              class="chakra-form-control css-1u79ci6"
              role="group"
            >
              <label
                class="chakra-form__label css-t3kal5"
                for="category"
                id="field-:r0:-label"
              >
                Category
              </label>
              <div
                class="chakra-select__wrapper css-42b2qy"
              >
                <select
                  class="chakra-select css-1ik61og"
                  id="category"
                  name="category"
                >
                  <option
                    value="9"
                  >
                    General Knowledge
                  </option>
                  <option
                    value="10"
                  >
                    Books
                  </option>
                  <option
                    value="14"
                  >
                    Television
                  </option>
                  <option
                    value="19"
                  >
                    Mathematics
                  </option>
                  <option
                    value="31"
                  >
                    Japanese Anime & Manga
                  </option>
                </select>
                <div
                  class="chakra-select__icon-wrapper css-162mkon"
                >
                  <svg
                    aria-hidden="true"
                    class="chakra-select__icon"
                    focusable="false"
                    role="presentation"
                    style="width: 1em; height: 1em; color: currentColor;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="chakra-form-control css-1u79ci6"
              role="group"
            >
              <label
                class="chakra-form__label css-t3kal5"
                for="difficulty"
                id="field-:r1:-label"
              >
                Difficulty
              </label>
              <div
                class="chakra-select__wrapper css-42b2qy"
              >
                <select
                  class="chakra-select css-1ik61og"
                  id="difficulty"
                  name="difficulty"
                >
                  <option
                    value="any"
                  >
                    Any
                  </option>
                  <option
                    value="easy"
                  >
                    Easy
                  </option>
                  <option
                    value="medium"
                  >
                    Medium
                  </option>
                  <option
                    value="hard"
                  >
                    Hard
                  </option>
                </select>
                <div
                  class="chakra-select__icon-wrapper css-162mkon"
                >
                  <svg
                    aria-hidden="true"
                    class="chakra-select__icon"
                    focusable="false"
                    role="presentation"
                    style="width: 1em; height: 1em; color: currentColor;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="chakra-form-control css-1u79ci6"
              role="group"
            >
              <label
                class="chakra-form__label css-19h9qxr"
                for="amount"
                id="field-:r2:-label"
              >
                Number of questions
              </label>
              <div
                class="chakra-select__wrapper css-42b2qy"
              >
                <select
                  class="chakra-select css-1ik61og"
                  id="amount"
                  name="amount"
                >
                  <option
                    value="5"
                  >
                    5
                  </option>
                  <option
                    value="10"
                  >
                    10
                  </option>
                  <option
                    value="15"
                  >
                    15
                  </option>
                </select>
                <div
                  class="chakra-select__icon-wrapper css-162mkon"
                >
                  <svg
                    aria-hidden="true"
                    class="chakra-select__icon"
                    focusable="false"
                    role="presentation"
                    style="width: 1em; height: 1em; color: currentColor;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="chakra-form-control css-1u79ci6"
              role="group"
            >
              <label
                class="chakra-form__label css-19h9qxr"
                for="gameMode"
                id="field-:r3:-label"
              >
                Game mode
              </label>
              <div
                class="chakra-select__wrapper css-42b2qy"
              >
                <select
                  class="chakra-select css-1ik61og"
                  id="gameMode"
                  name="gameMode"
                >
                  <option
                    value="single"
                  >
                    Single player
                  </option>
                  <option
                    value="multi"
                  >
                    Multiplayer
                  </option>
                </select>
                <div
                  class="chakra-select__icon-wrapper css-162mkon"
                >
                  <svg
                    aria-hidden="true"
                    class="chakra-select__icon"
                    focusable="false"
                    role="presentation"
                    style="width: 1em; height: 1em; color: currentColor;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <button
            class="chakra-button css-1w54gy2"
            type="submit"
          >
            Start trivia
          </button>
        </form>
      </div>
    </div>
  </main>
  <footer
    class="css-1s2lyti"
  >
    <div
      class="chakra-container css-0"
    >
      <div
        class="chakra-stack css-11h4xmw"
      >
        <a
          href="/"
        >
          <p
            class="chakra-text css-rekzoj"
          >
            Trivia
          </p>
        </a>
        <div
          class="css-0"
        >
          <strong
            class="chakra-heading css-kf3i1p"
          >
            Site map
          </strong>
          <div
            class="chakra-stack css-1ko7mmx"
          >
            <a
              href="/"
            >
              Home
            </a>
            <a
              href="/login"
            >
              Sign in
            </a>
            <a
              href="/sign-up"
            >
              Sign up
            </a>
          </div>
        </div>
        <div
          class="css-0"
        >
          <strong
            class="chakra-heading css-kf3i1p"
          >
            Social
          </strong>
          <div
            class="chakra-stack css-1ko7mmx"
          >
            <a
              href="https://www.linkedin.com/in/ingrid-flack/"
            >
              LinkedIn
            </a>
            <a
              href="https://github.com/ingridflack"
            >
              GitHub
            </a>
          </div>
        </div>
      </div>
      <div
        class="chakra-stack css-1y3yl0q"
      >
        <p
          class="chakra-text css-0"
        >
          © 2024 Trivia App
        </p>
        <p
          class="chakra-text css-0"
        >
          Developed by Ingrid Flack
        </p>
      </div>
    </div>
  </footer>
</div>
`;
